### ISSUE TITLE
Stack buffer overflows don't throw exception when stack tagging enabled  with arm64 MTE

### ISSUE THREAD

**Issue by aditi741997** [llvm/llvm-project#54192]
I am compiling the following code [test_mte_stack2_code.txt](https://github.com/llvm/llvm-project/files/8183382/test_mte_stack2_code.txt)] using 'clang-11 -target aarch64-linux-gnu -march=armv8.5-a+memtag -fsanitize=memtag -I/usr/aarch64-linux-gnu/include/c++/9/aarch64-linux-gnu test_mte_stack2.cpp -static'.

I run the executable on a VM in QEMU, with MTE on [I ran the example code in ? and got a segfault, meaning QEMU, Kernel are behaving correctly]

I observed that 
1. local variables in main() are not getting tagged
2. In test function, local variables are being tagged but I am able to do stack buffer overflow from arr2 into arr1, without modifying the tag. 
3. Similarly in test2, which gets the pointer to one of test's local variables (a), i can read (i.e. overflow into) test's other local variables using the pointer to a.

What are the reasons behind the above? Shouldn't the tag be checked for accesses in functions test, test2 to detect overflow?

I have attached the assembly file [test_mte_stack2_assembly.txt](https://github.com/llvm/llvm-project/files/8183416/test_mte_stack2_assembly.txt)
 [which has no ldg instructions] and the output of running the executable in the VM 
[test_mte_stack2_output.txt](https://github.com/llvm/llvm-project/files/8183419/test_mte_stack2_output.txt).

**Comment by efriedma-quic** [comment]
Yes, this is how stack tagging currently works.  From https://llvm.org/docs/MemTagSanitizer.html#stack-instrumentation : "[...] MemTagSanitizer generates at most one random tag per function [...]".  There might be room for improvement.

**Comment by john-brawn-arm** [comment]
If I take this example and run it in a bare-metal environment on an Arm FastModel then I get an exception on the access to arr[ind] in main, so my guess is that this is Linux or QEMU problem rather than a clang problem.

Specifically looking at the disassembly I see at the start of main that it's tagging the array arr and spilling the tagged pointer to the stack
```
	addg	x15, x8, #80, #4
	stg	x15, [x15, #32]
	st2g	x15, [x15]
	str	x15, [sp, #32]                  // 8-byte Folded Spill
```
then later it reloads the tagged pointer and the read arr[ind] is done through it
```
	ldr	x10, [sp, #32]                  // 8-byte Folded Reload
	ldr	w1, [x10, x9, lsl #2]
```
meaning that in the load the tag in x10 should be compared against the tag of the memory that's accessed causing an exception if they don't match.

**Comment by aditi741997** [comment]
It turned out that I needed to specifically set the flag PROT_MTE for stack pages. I'm not sure why it behaved as expected on Arm fastModel?
